matrix_functions.py

```diff
@@ -200,15 +203,37 @@ def matrix_eigendecomposition(
             device=A.device,
         )
 
+    # TODO: reduce redundant code when noninvertible_handling_config is generalized to all methods
+    # check epsilon is 0 when using pseudo-inverse
+    if (
+        isinstance(
+            eigendecomposition_config.noninvertible_handling_config,
+            PseudoInverseConfig,
+        )
+        and epsilon != 0.0
+    ):
+        raise ValueError(f"{epsilon=} should be 0.0 when using pseudo-inverse!")
+
+    # Add epsilon to the diagonal to help with numerical stability of the eigenvalue decomposition
+    # Only do it when add_epsilon_before_computation is True (root_inv_config must be a RegularizationConfig)
+    if getattr(eigendecomposition_config, "add_epsilon_before_computation", False):
```

Comment:
Shouldn't it be `eigendecomposition_config.noninvertible_handling_config`? I.e.:
```suggestion
    # Only do it when add_epsilon_before_computation is True (eigendecomposition_config.noninvertible_handling_config must be a RegularizationConfig)
    if getattr(eigendecomposition_config.noninvertible_handling_config, "add_epsilon_before_computation", False):
```
If this fix is correct it means that before `getattr` always returned `False` and there is no test that catches this.

Reply:
Good catch! I'll definitely need to improve testing, because this is exactly the kind of situation everyone is worried about...